File: gherkin/javascript/lib/Gherkin.js
const {spawn} = require('child_process')
const ExeFile = require('./exe/ExeFile')
const cm = require('cucumber-messages').io.cucumber.messages
const ProtobufMessageStream = require('./ProtobufMessageStream')

module.exports = class Gherkin {
  constructor(paths, sources, options) {
    this._paths = paths
    this._sources = sources
    this._options = Object.assign({includeSource: true, includeGherkinDocument: true, includePickles: true}, options)
    this._exeFile = new ExeFile(`${__dirname}/../gherkin-go/gherkin-go-{{.OS}}-{{.Arch}}{{.Ext}}`)
  }

  messages() {
    const options = []
    if (!this._options.includeSource) options.push('--no-source')
    if (!this._options.includeGherkinDocument) options.push('--no-ast')
    if (!this._options.includePickles) options.push('--no-pickles')
    const args = options.concat(this._paths)
    const gherkin = spawn(this._exeFile.fileName, args)
    for (const source of this._sources) {
      gherkin.stdin.write(cm.Source.encodeDelimited(source).finish())
    }
    gherkin.stdin.end()
    return gherkin.stdout.pipe(new ProtobufMessageStream(cm.Wrapper))
  }
}

File: gherkin/dart/bin/gherkin.dart
import 'dart:io';

import 'package:gherkin/Gherkin.dart';
import 'package:gherkin/exceptions.dart';
import 'package:gherkin/language.dart';
import 'package:gherkin/messages.dart';

void main(List<String> args) async
{
  var includeSource = true;
  var includeAst = true;
  var includePickles = true;
  var paths = <String>[];
  var idGenerator = IdGenerator.uuidGenerator;

  for( var arg in args ) {
    switch(arg) {
      case '--no-source':
        includeSource = false;
        break;
      case '--no-ast':
        includeAst = false;
        break;
      case '--no-pickles':
        includePickles = false;
        break;
      case '--predictable-ids':
        idGenerator = IdGenerator.incrementingGenerator;
        break;
      default:
        paths.add(arg);
    }

    var messageWriter = makeMessageWriter();

    var messages = Gherkin.fromPaths(paths, includeSource
        , includeAst, includePickles, idGenerator);
    printMessages(messageWriter, messages);
  }
}

IMessageWriter makeMessageWriter() {
  return MessageToNdjsonWriter(stdout);
}

void printMessages(IMessageWriter messageWriter, Stream<Envelope> messages) {
  messages.forEach((envelope) {
    try {
      messageWriter.write(envelope);
    }
    on IOException catch (e) {
      throw GherkinException("Couldn't print messages", e);
    }
  });
}

